File: dlang/psi-impl/src/main/java/io/github/intellij/dlanguage/stubs/types/DestructorStubElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import io.github.intellij.dlanguage.psi.impl.DLanguageDestructorImpl;
import io.github.intellij.dlanguage.psi.named.DlangDestructor;
import io.github.intellij.dlanguage.stubs.DlangDestructorStub;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class DestructorStubElementType extends
    DStubElementType<DlangDestructorStub, DlangDestructor> {
    public DestructorStubElementType(final String debugName) {
        super(debugName);
    }

    @Override
    public DlangDestructor createPsi(@NotNull final DlangDestructorStub stub) {
        return new DLanguageDestructorImpl(stub, this);
    }

    @NotNull
    @Override
    public DlangDestructorStub createStub(@NotNull final DlangDestructor psi,
        final StubElement parentStub) {
        return new DlangDestructorStub(parentStub, this);
    }

    @NotNull
    @Override
    public DlangDestructorStub deserialize(@NotNull final StubInputStream dataStream, final StubElement parentStub) throws IOException {
        return new DlangDestructorStub(parentStub, this);
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/DestructorStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import io.github.intellij.dlanguage.psi.impl.DLanguageDestructorImpl
import io.github.intellij.dlanguage.psi.named.DLanguageDestructor
import io.github.intellij.dlanguage.stubs.DLanguageDestructorStub
import java.io.IOException

class DestructorStubElementType(debugName: String) : DStubElementType<DLanguageDestructorStub, DLanguageDestructor>(debugName) {
    override fun createPsi(stub: DLanguageDestructorStub): DLanguageDestructor {
        return DLanguageDestructorImpl(stub, this)
    }

    override fun createStub(
        psi: DLanguageDestructor,
        parentStub: StubElement<*>?
    ): DLanguageDestructorStub {
        return DLanguageDestructorStub(parentStub, this)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>?): DLanguageDestructorStub {
        return DLanguageDestructorStub(parentStub, this)
    }
}
